src/NLog/Targets/FileArchiveStyleFactory.cs

```diff
@@ -0,0 +1,758 @@
+        /// <returns></returns>
+        IEnumerable<DateAndSequenceArchive> CheckArchiveCleanup(string archiveFilePath, List<DateAndSequenceArchive> existingArchiveFiles, int maxArchiveFiles);
+    };
+
+    static class FileArchiveStyleFactory
+    {
+        public static IFileArchiveStyle CreateArchiveStyle(string archiveFilePath, ArchiveNumberingMode archiveNumbering, string dateFormat)
+        {
+            if (ContainsFileNamePattern(archiveFilePath))
+            {
+                switch (archiveNumbering)
+                {
+                    case ArchiveNumberingMode.Rolling: return new FileArchiveModeRolling();
+                    case ArchiveNumberingMode.Sequence: return new FileArchiveModeSequence(dateFormat);
+                    case ArchiveNumberingMode.Date: return new FileArchiveModeDate(dateFormat);
+                    case ArchiveNumberingMode.DateAndSequence: return new FileArchiveModeDateAndSequence(dateFormat);
+                }
+            }
+
+            return new FileArchiveModeDynamicSequence(archiveNumbering, dateFormat);
+        }
+
+        /// <summary>
+        /// Determines if the file name as <see cref="String"/> contains a numeric pattern i.e. {#} in it.  
+        ///
+        /// Example: 
+        ///     trace{#}.log        Contains the numeric pattern.
+        ///     trace{###}.log      Contains the numeric pattern.
+        ///     trace{#X#}.log      Contains the numeric pattern (See remarks).
+        ///     trace.log           Does not contain the pattern.
+        /// </summary>
+        /// <remarks>Occasionally, this method can identify the existence of the {#} pattern incorrectly.</remarks>
+        /// <param name="fileName">File name to be checked.</param>
+        /// <returns><see langword="true"/> when the pattern is found; <see langword="false"/> otherwise.</returns>
+        public static bool ContainsFileNamePattern(string fileName)
+        {
+            int startingIndex = fileName.IndexOf("{#", StringComparison.Ordinal);
+            int endingIndex = fileName.IndexOf("#}", StringComparison.Ordinal);
+
+            return (startingIndex != -1 && endingIndex != -1 && startingIndex < endingIndex);
+        }
+    }
+
+    sealed class FileNameTemplate
+    {
+        /// <summary>
+        /// Characters determining the start of the <see cref="P:FileNameTemplate.Pattern"/>.
+        /// </summary>
+        public const string PatternStartCharacters = "{#";
+
+        /// <summary>
+        /// Characters determining the end of the <see cref="P:FileNameTemplate.Pattern"/>.
+        /// </summary>
+        public const string PatternEndCharacters = "#}";
+
+        /// <summary>
+        /// File name which is used as template for matching and replacements. 
+        /// It is expected to contain a pattern to match.
+        /// </summary>
+        public string Template { get { return this.template; } }
```

Comment:
Would be great if you would do that as I cannot see what's new or what's moved (easily). Thx in advance!